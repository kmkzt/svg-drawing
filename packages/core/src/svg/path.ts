import { toAbsoluteCommands, toRelativeCommands } from './command'
import type {
  PathClass,
  CommandClass,
  PathAttributes,
  PointObject,
  PathObject,
} from '../types'

/**
 * Cannot support commands that use `M` or` z` more than once.
 *
 * Not support example: `M 0 0 L 1 1 Z M 1 1 L 2 2 Z`.
 */
export class Path implements PathClass {
  private static index = 0
  private static genIndex = () => (Path.index += 1)
  private _relativeCommands: CommandClass[] = []
  public key: string
  constructor(public attrs: PathAttributes = {}, _key?: string) {
    this.key = _key || `p${Path.genIndex()}`
  }

  get absoluteCommands(): CommandClass[] {
    return toAbsoluteCommands(this._relativeCommands)
  }

  get relativeCommands(): CommandClass[] {
    return this._relativeCommands
  }

  setCommands(commands: ReadonlyArray<CommandClass> | Readonly<CommandClass>) {
    this._relativeCommands = toRelativeCommands([commands].flat())

    return this
  }

  addCommand(command: Readonly<CommandClass>) {
    this.setCommands([...this._relativeCommands, command])

    return this
  }

  updateCommand(
    i: number,
    update: (absoluteCommand: CommandClass) => CommandClass
  ) {
    const commands = this.absoluteCommands
    commands[i] = update(commands[i])

    this.setCommands(commands)
    return this
  }

  deleteCommand(i: number) {
    this.setCommands(this.absoluteCommands.filter((_, index) => index !== i))
    return this
  }

  scale(r: number) {
    this.setCommands(
      this._relativeCommands.map((c: CommandClass) => c.scale(r))
    )
    return this
  }

  scaleX(r: number) {
    this.setCommands(
      this._relativeCommands.map((c: CommandClass) => c.scaleX(r))
    )
    return this
  }

  scaleY(r: number) {
    this.setCommands(
      this._relativeCommands.map((c: CommandClass) => c.scaleY(r))
    )
    return this
  }

  getCommandString() {
    return (
      this.relativeCommands
        ?.map((com: CommandClass, _i: number) => com.toString())
        .join(' ')
        .trim() || ''
    )
  }

  toJson(): PathObject {
    return {
      key: this.key,
      type: 'path',
      attributes: {
        ...this.attrs,
        d: this.getCommandString(),
      },
    }
  }

  setAttributes(attrs: PathAttributes) {
    this.attrs = attrs

    return this
  }

  updateAttributes(attrs: PathAttributes) {
    this.attrs = {
      ...this.attrs,
      ...attrs,
    }

    return this
  }

  translate(po: PointObject) {
    if (
      this._relativeCommands.length === 0 &&
      this._relativeCommands[0].type !== 'M'
    ) {
      return this
    }

    this._relativeCommands[0] = this._relativeCommands[0].translate(po)

    return this
  }

  clone() {
    return new Path(
      {
        ...this.attrs,
      },
      this.key
    ).setCommands(this._relativeCommands.map((c) => c.clone()))
  }
}
