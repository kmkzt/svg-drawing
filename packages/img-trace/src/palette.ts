import { convertRGBAImage } from './utils/convertRGBAImage'

export interface Rgba {
  r: number
  g: number
  b: number
  a: number
}

const DEFAULT_NUMBER_OF_COLORS = 16
export interface FromImageDataOptions {
  numberOfColors?: number
  colorQuantCycles?: number
}
export class Palette {
  // Deterministic sampling a palette from imagedata: rectangular grid
  public static imageData(
    argimgd: ImageData,
    { numberOfColors, colorQuantCycles }: FromImageDataOptions = {}
  ): Rgba[] {
    const nc = numberOfColors || DEFAULT_NUMBER_OF_COLORS
    const cqc = colorQuantCycles || 3
    const imgd = convertRGBAImage(argimgd)

    const palette = this._deterministic(imgd, nc)
    let paletteacc: {
      r: number
      g: number
      b: number
      a: number
      n: number
    }[] = []
    // Using a form of k-means clustering repeatead options.colorquantcycles times. http://en.wikipedia.org/wiki/Color_quantization
    for (let cnt = 0; cnt < cqc; cnt++) {
      if (cnt > 0) {
        for (let k = 0; k < palette.length; k++) {
          // averaging
          if (paletteacc[k].n > 0) {
            palette[k] = {
              r: Math.floor(paletteacc[k].r / paletteacc[k].n),
              g: Math.floor(paletteacc[k].g / paletteacc[k].n),
              b: Math.floor(paletteacc[k].b / paletteacc[k].n),
              a: Math.floor(paletteacc[k].a / paletteacc[k].n),
            }
          }
        }
      }

      paletteacc = Array.from({ length: palette.length }, () => ({
        r: 0,
        g: 0,
        b: 0,
        a: 0,
        n: 0,
      }))

      // loop through all pixels
      for (let j = 0; j < imgd.height; j++) {
        for (let i = 0; i < imgd.width; i++) {
          // pixel index
          const idx = (j * imgd.width + i) * 4

          // find closest color from palette by measuring (rectilinear) color distance between this pixel and all palette colors
          // In my experience, https://en.wikipedia.org/wiki/Rectilinear_distance works better than https://en.wikipedia.org/wiki/Euclidean_distance
          let cdl = 1024 // 4 * 256 is the maximum RGBA distance
          const ci = palette.reduce((findId: number, pal: Rgba, id: number) => {
            const cd =
              Math.abs(pal.r - imgd.data[idx]) +
              Math.abs(pal.g - imgd.data[idx + 1]) +
              Math.abs(pal.b - imgd.data[idx + 2]) +
              Math.abs(pal.a - imgd.data[idx + 3])
            if (cd < cdl) {
              cdl = cd
              return id
            }
            return findId
          }, 0)

          // add to palettacc
          paletteacc[ci].r += imgd.data[idx]
          paletteacc[ci].g += imgd.data[idx + 1]
          paletteacc[ci].b += imgd.data[idx + 2]
          paletteacc[ci].a += imgd.data[idx + 3]
          paletteacc[ci].n += 1
        }
      }
    }
    return palette
  }

  // Deterministic sampling a palette from imagedata: rectangular grid
  private static _deterministic(
    imgd: ImageData,
    numberOfColors: number
  ): Rgba[] {
    const palette: Rgba[] = []
    const ni = Math.ceil(Math.sqrt(numberOfColors))
    const nj = Math.ceil(numberOfColors / ni)
    const vx = imgd.width / (ni + 1)
    const vy = imgd.height / (nj + 1)
    for (let j = 0; j < nj; j++) {
      for (let i = 0; i < ni; i++) {
        if (palette.length === numberOfColors) break
        const idx = Math.floor((j + 1) * vy * imgd.width + (i + 1) * vx) * 4
        palette.push({
          r: imgd.data[idx],
          g: imgd.data[idx + 1],
          b: imgd.data[idx + 2],
          a: imgd.data[idx + 3],
        })
      }
    }
    return palette
  }
  // Generating a palette with number of colors
  public static number(
    numberofcolors: number = DEFAULT_NUMBER_OF_COLORS
  ): Rgba[] {
    if (numberofcolors < 8) return this.grey(numberofcolors)
    const palette: Rgba[] = []
    const colorqnum = Math.floor(Math.pow(numberofcolors, 1 / 3))
    const colorstep = Math.floor(255 / (colorqnum - 1))

    for (let r = 0; r < colorqnum; r += 1) {
      for (let g = 0; g < colorqnum; g += 1) {
        for (let b = 0; b < colorqnum; b += 1) {
          palette.push({
            r: r * colorstep,
            g: g * colorstep,
            b: b * colorstep,
            a: 255,
          })
        }
      }
    }

    // TODO: rest improve
    // TODO: test
    const rest = numberofcolors - colorqnum * colorqnum * colorqnum
    for (let rcnt = 0; rcnt < rest; rcnt++) {
      palette.push({
        r: Math.floor(Math.random() * 255),
        g: Math.floor(Math.random() * 255),
        b: Math.floor(Math.random() * 255),
        a: Math.floor(Math.random() * 255),
      })
    }

    return palette
  }

  public static grey(
    numberofcolors: number = DEFAULT_NUMBER_OF_COLORS
  ): Rgba[] {
    const palette: Rgba[] = []
    const graystep = Math.floor(255 / (numberofcolors - 1))
    for (let i = 0; i < numberofcolors; i++) {
      palette.push({
        r: i * graystep,
        g: i * graystep,
        b: i * graystep,
        a: 255,
      })
    }

    return palette
  }
}
