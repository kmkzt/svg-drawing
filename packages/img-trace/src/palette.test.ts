import { Palette, FromImageDataOptions, Rgba } from './palette'
import { PNGReader } from './PNGReader'
import { readFile } from 'fs'
import { resolve } from 'path'
interface PngData {
  width: number
  height: number
  data: any
}
const loadPngData = (filepath: string, cb: (png: PngData) => void) => {
  readFile(
    filepath, // Input file path
    (err: any, bytes: any) => {
      if (err) {
        console.log(err)
        throw err
      }

      const reader = new PNGReader(bytes)

      reader.parse(
        (err, png) => {
          if (err) {
            console.log(err)
            throw err
          }

          // tracing to SVG string
          cb({
            width: png.width,
            height: png.height,
            data: png.pixels,
          })
        } // End of readFile callback()
      ) // End of fs.readFile()
    }
  )
}

const TEST_NUMBER_OF_COLORS = [undefined, 2, 4, 7, 8, 27, 64]
const TEST_COLOR_QUANT_CYCLES = [undefined, 1, 8]
const getTestResult = (pal: Rgba[]) =>
  pal.map((c) => `rgba(${c.r}, ${c.g}, ${c.b}, ${c.a / 255})`)
describe('palette.ts', () => {
  const testimage = resolve(__dirname, '__testdata__/panda.png')
  describe('Palette', () => {
    describe('Palette.imageData', () => {
      TEST_COLOR_QUANT_CYCLES.map((colorQuantCycles: number | undefined) => {
        TEST_NUMBER_OF_COLORS.map((numberOfColors: number | undefined) => {
          it(`colorQuantCycles: ${
            colorQuantCycles || 'default'
          } ;numberOfColors: ${numberOfColors || 'default'}`, (done) => {
            loadPngData(testimage, (imgd: PngData) => {
              expect(
                getTestResult(
                  Palette.imageData(imgd, { colorQuantCycles, numberOfColors })
                )
              ).toMatchSnapshot()
              done()
            })
          })
        })
      })
    })
    describe('Palette.number', () => {
      // TODO: Rest colors
      ;[8, 27].map((arg: number | undefined) => {
        it(`numberOfColors: ${arg || 'default'}`, () => {
          expect(getTestResult(Palette.number(arg))).toMatchSnapshot()
        })
      })
    })
    describe('Palette.gray', () => {
      ;[2, 7, 16].map((arg: number | undefined) => {
        it(`numberOfColors: ${arg || 'default'}`, () => {
          expect(getTestResult(Palette.grey(arg))).toMatchSnapshot()
        })
      })
    })
  })
})
