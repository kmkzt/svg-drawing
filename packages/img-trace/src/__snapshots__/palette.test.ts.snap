// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`palette.ts Palette Palette.gray numberOfColors: 2 1`] = `
Array [
  "rgba(0, 0, 0, 1)",
  "rgba(255, 255, 255, 1)",
]
`;

exports[`palette.ts Palette Palette.gray numberOfColors: 7 1`] = `
Array [
  "rgba(0, 0, 0, 1)",
  "rgba(42, 42, 42, 1)",
  "rgba(84, 84, 84, 1)",
  "rgba(126, 126, 126, 1)",
  "rgba(168, 168, 168, 1)",
  "rgba(210, 210, 210, 1)",
  "rgba(252, 252, 252, 1)",
]
`;

exports[`palette.ts Palette Palette.gray numberOfColors: 16 1`] = `
Array [
  "rgba(0, 0, 0, 1)",
  "rgba(17, 17, 17, 1)",
  "rgba(34, 34, 34, 1)",
  "rgba(51, 51, 51, 1)",
  "rgba(68, 68, 68, 1)",
  "rgba(85, 85, 85, 1)",
  "rgba(102, 102, 102, 1)",
  "rgba(119, 119, 119, 1)",
  "rgba(136, 136, 136, 1)",
  "rgba(153, 153, 153, 1)",
  "rgba(170, 170, 170, 1)",
  "rgba(187, 187, 187, 1)",
  "rgba(204, 204, 204, 1)",
  "rgba(221, 221, 221, 1)",
  "rgba(238, 238, 238, 1)",
  "rgba(255, 255, 255, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 1 ;numberOfColors: 2 1`] = `
Array [
  "rgba(226, 211, 193, 1)",
  "rgba(20, 67, 104, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 1 ;numberOfColors: 4 1`] = `
Array [
  "rgba(127, 153, 99, 1)",
  "rgba(221, 213, 206, 1)",
  "rgba(225, 222, 223, 1)",
  "rgba(15, 17, 38, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 1 ;numberOfColors: 7 1`] = `
Array [
  "rgba(124, 153, 103, 1)",
  "rgba(128, 150, 95, 1)",
  "rgba(220, 214, 216, 1)",
  "rgba(135, 156, 103, 1)",
  "rgba(224, 219, 210, 1)",
  "rgba(9, 62, 102, 1)",
  "rgba(225, 222, 222, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 1 ;numberOfColors: 8 1`] = `
Array [
  "rgba(124, 153, 103, 1)",
  "rgba(128, 150, 95, 1)",
  "rgba(220, 214, 216, 1)",
  "rgba(135, 156, 103, 1)",
  "rgba(224, 219, 210, 1)",
  "rgba(9, 62, 102, 1)",
  "rgba(225, 222, 222, 1)",
  "rgba(232, 241, 235, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 1 ;numberOfColors: 27 1`] = `
Array [
  "rgba(122, 148, 96, 1)",
  "rgba(123, 149, 100, 1)",
  "rgba(6, 58, 100, 1)",
  "rgba(220, 203, 188, 1)",
  "rgba(216, 199, 182, 1)",
  "rgba(12, 45, 77, 1)",
  "rgba(19, 61, 104, 1)",
  "rgba(128, 154, 104, 1)",
  "rgba(129, 152, 96, 1)",
  "rgba(221, 203, 184, 1)",
  "rgba(219, 203, 186, 1)",
  "rgba(10, 61, 100, 1)",
  "rgba(241, 251, 248, 1)",
  "rgba(134, 157, 100, 1)",
  "rgba(226, 209, 189, 1)",
  "rgba(222, 208, 192, 1)",
  "rgba(20, 68, 106, 1)",
  "rgba(25, 72, 108, 1)",
  "rgba(230, 213, 197, 1)",
  "rgba(229, 211, 193, 1)",
  "rgba(223, 207, 192, 1)",
  "rgba(17, 16, 34, 1)",
  "rgba(14, 18, 37, 1)",
  "rgba(235, 244, 236, 1)",
  "rgba(232, 216, 193, 1)",
  "rgba(223, 208, 192, 1)",
  "rgba(231, 242, 235, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 1 ;numberOfColors: 64 1`] = `
Array [
  "rgba(214, 196, 181, 1)",
  "rgba(216, 222, 212, 1)",
  "rgba(116, 143, 91, 1)",
  "rgba(119, 145, 93, 1)",
  "rgba(121, 147, 96, 1)",
  "rgba(122, 147, 99, 1)",
  "rgba(121, 148, 98, 1)",
  "rgba(14, 59, 102, 1)",
  "rgba(222, 206, 191, 1)",
  "rgba(118, 143, 85, 1)",
  "rgba(217, 200, 182, 1)",
  "rgba(220, 216, 210, 1)",
  "rgba(18, 61, 99, 1)",
  "rgba(120, 148, 96, 1)",
  "rgba(123, 150, 98, 1)",
  "rgba(125, 152, 101, 1)",
  "rgba(16, 70, 103, 1)",
  "rgba(17, 68, 106, 1)",
  "rgba(127, 153, 99, 1)",
  "rgba(129, 152, 95, 1)",
  "rgba(221, 204, 183, 1)",
  "rgba(221, 213, 206, 1)",
  "rgba(17, 63, 102, 1)",
  "rgba(121, 149, 100, 1)",
  "rgba(18, 67, 106, 1)",
  "rgba(20, 70, 106, 1)",
  "rgba(20, 71, 109, 1)",
  "rgba(201, 119, 109, 1)",
  "rgba(115, 93, 100, 1)",
  "rgba(133, 155, 98, 1)",
  "rgba(225, 208, 189, 1)",
  "rgba(223, 223, 221, 1)",
  "rgba(224, 220, 211, 1)",
  "rgba(222, 207, 192, 1)",
  "rgba(20, 68, 108, 1)",
  "rgba(233, 243, 238, 1)",
  "rgba(235, 244, 238, 1)",
  "rgba(18, 13, 20, 1)",
  "rgba(237, 242, 236, 1)",
  "rgba(139, 153, 97, 1)",
  "rgba(227, 212, 197, 1)",
  "rgba(230, 213, 195, 1)",
  "rgba(225, 222, 223, 1)",
  "rgba(223, 207, 191, 1)",
  "rgba(228, 224, 231, 1)",
  "rgba(15, 17, 38, 1)",
  "rgba(234, 242, 236, 1)",
  "rgba(235, 244, 236, 1)",
  "rgba(236, 243, 235, 1)",
  "rgba(235, 243, 234, 1)",
  "rgba(229, 213, 197, 1)",
  "rgba(230, 213, 195, 1)",
  "rgba(225, 223, 223, 1)",
  "rgba(224, 206, 192, 1)",
  "rgba(231, 239, 235, 1)",
  "rgba(17, 20, 42, 1)",
  "rgba(233, 243, 236, 1)",
  "rgba(18, 22, 47, 1)",
  "rgba(15, 19, 44, 1)",
  "rgba(19, 20, 41, 1)",
  "rgba(230, 215, 198, 1)",
  "rgba(230, 213, 195, 1)",
  "rgba(224, 215, 211, 1)",
  "rgba(224, 207, 191, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 1 ;numberOfColors: default 1`] = `
Array [
  "rgba(219, 207, 213, 1)",
  "rgba(121, 146, 96, 1)",
  "rgba(124, 151, 100, 1)",
  "rgba(18, 65, 102, 1)",
  "rgba(131, 154, 97, 1)",
  "rgba(224, 223, 221, 1)",
  "rgba(83, 116, 107, 1)",
  "rgba(134, 160, 108, 1)",
  "rgba(235, 244, 236, 1)",
  "rgba(229, 213, 196, 1)",
  "rgba(229, 225, 226, 1)",
  "rgba(21, 67, 101, 1)",
  "rgba(15, 19, 44, 1)",
  "rgba(249, 255, 249, 1)",
  "rgba(231, 214, 197, 1)",
  "rgba(232, 231, 225, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 8 ;numberOfColors: 2 1`] = `
Array [
  "rgba(226, 217, 204, 1)",
  "rgba(71, 92, 85, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 8 ;numberOfColors: 4 1`] = `
Array [
  "rgba(130, 145, 99, 1)",
  "rgba(223, 208, 192, 1)",
  "rgba(233, 240, 233, 1)",
  "rgba(18, 44, 72, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 8 ;numberOfColors: 7 1`] = `
Array [
  "rgba(124, 149, 98, 1)",
  "rgba(89, 109, 95, 1)",
  "rgba(215, 199, 181, 1)",
  "rgba(204, 123, 109, 1)",
  "rgba(226, 211, 195, 1)",
  "rgba(18, 43, 72, 1)",
  "rgba(233, 241, 233, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 8 ;numberOfColors: 8 1`] = `
Array [
  "rgba(124, 149, 98, 1)",
  "rgba(89, 109, 95, 1)",
  "rgba(214, 199, 181, 1)",
  "rgba(204, 123, 109, 1)",
  "rgba(226, 210, 193, 1)",
  "rgba(18, 43, 72, 1)",
  "rgba(224, 222, 216, 1)",
  "rgba(234, 242, 235, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 8 ;numberOfColors: 27 1`] = `
Array [
  "rgba(115, 140, 91, 1)",
  "rgba(121, 147, 96, 1)",
  "rgba(27, 48, 79, 1)",
  "rgba(215, 199, 183, 1)",
  "rgba(186, 175, 153, 1)",
  "rgba(63, 61, 65, 1)",
  "rgba(16, 61, 99, 1)",
  "rgba(205, 122, 108, 1)",
  "rgba(128, 152, 99, 1)",
  "rgba(219, 203, 181, 1)",
  "rgba(212, 197, 178, 1)",
  "rgba(13, 56, 92, 1)",
  "rgba(246, 252, 247, 1)",
  "rgba(151, 145, 130, 1)",
  "rgba(223, 207, 187, 1)",
  "rgba(211, 207, 204, 1)",
  "rgba(19, 68, 105, 1)",
  "rgba(77, 100, 99, 1)",
  "rgba(220, 218, 214, 1)",
  "rgba(226, 210, 193, 1)",
  "rgba(219, 203, 187, 1)",
  "rgba(16, 11, 19, 1)",
  "rgba(17, 19, 40, 1)",
  "rgba(233, 242, 235, 1)",
  "rgba(230, 214, 196, 1)",
  "rgba(222, 207, 191, 1)",
  "rgba(228, 227, 221, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 8 ;numberOfColors: 64 1`] = `
Array [
  "rgba(189, 177, 156, 1)",
  "rgba(216, 217, 211, 1)",
  "rgba(88, 122, 102, 1)",
  "rgba(113, 139, 87, 1)",
  "rgba(117, 143, 92, 1)",
  "rgba(128, 118, 120, 1)",
  "rgba(120, 146, 94, 1)",
  "rgba(9, 51, 88, 1)",
  "rgba(217, 201, 185, 1)",
  "rgba(115, 94, 86, 1)",
  "rgba(212, 197, 178, 1)",
  "rgba(212, 205, 207, 1)",
  "rgba(35, 49, 78, 1)",
  "rgba(104, 143, 102, 1)",
  "rgba(121, 148, 98, 1)",
  "rgba(124, 152, 101, 1)",
  "rgba(15, 60, 100, 1)",
  "rgba(17, 65, 103, 1)",
  "rgba(129, 152, 96, 1)",
  "rgba(126, 149, 92, 1)",
  "rgba(216, 199, 181, 1)",
  "rgba(204, 207, 196, 1)",
  "rgba(15, 58, 94, 1)",
  "rgba(119, 145, 98, 1)",
  "rgba(21, 64, 97, 1)",
  "rgba(50, 74, 89, 1)",
  "rgba(21, 70, 106, 1)",
  "rgba(205, 122, 108, 1)",
  "rgba(70, 97, 100, 1)",
  "rgba(134, 156, 102, 1)",
  "rgba(224, 208, 187, 1)",
  "rgba(220, 220, 215, 1)",
  "rgba(237, 223, 206, 1)",
  "rgba(221, 207, 191, 1)",
  "rgba(17, 67, 106, 1)",
  "rgba(234, 243, 236, 1)",
  "rgba(246, 253, 247, 1)",
  "rgba(10, 5, 14, 1)",
  "rgba(237, 242, 234, 1)",
  "rgba(162, 155, 137, 1)",
  "rgba(224, 208, 194, 1)",
  "rgba(227, 210, 191, 1)",
  "rgba(223, 218, 220, 1)",
  "rgba(221, 204, 188, 1)",
  "rgba(231, 228, 222, 1)",
  "rgba(17, 15, 29, 1)",
  "rgba(229, 239, 232, 1)",
  "rgba(231, 241, 234, 1)",
  "rgba(235, 242, 234, 1)",
  "rgba(230, 236, 228, 1)",
  "rgba(229, 213, 196, 1)",
  "rgba(230, 213, 193, 1)",
  "rgba(223, 227, 220, 1)",
  "rgba(223, 207, 191, 1)",
  "rgba(225, 231, 229, 1)",
  "rgba(17, 19, 43, 1)",
  "rgba(227, 239, 234, 1)",
  "rgba(66, 53, 47, 1)",
  "rgba(14, 17, 39, 1)",
  "rgba(26, 21, 31, 1)",
  "rgba(231, 215, 198, 1)",
  "rgba(230, 213, 195, 1)",
  "rgba(222, 214, 206, 1)",
  "rgba(221, 204, 183, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: 8 ;numberOfColors: default 1`] = `
Array [
  "rgba(213, 208, 204, 1)",
  "rgba(116, 142, 92, 1)",
  "rgba(124, 150, 98, 1)",
  "rgba(15, 60, 98, 1)",
  "rgba(140, 154, 111, 1)",
  "rgba(220, 219, 215, 1)",
  "rgba(79, 93, 92, 1)",
  "rgba(205, 123, 108, 1)",
  "rgba(233, 242, 235, 1)",
  "rgba(215, 199, 181, 1)",
  "rgba(234, 220, 204, 1)",
  "rgba(21, 67, 103, 1)",
  "rgba(17, 18, 37, 1)",
  "rgba(246, 253, 247, 1)",
  "rgba(226, 210, 193, 1)",
  "rgba(227, 228, 223, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: default ;numberOfColors: 2 1`] = `
Array [
  "rgba(226, 216, 203, 1)",
  "rgba(69, 92, 84, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: default ;numberOfColors: 4 1`] = `
Array [
  "rgba(130, 145, 99, 1)",
  "rgba(223, 208, 191, 1)",
  "rgba(233, 240, 233, 1)",
  "rgba(18, 44, 72, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: default ;numberOfColors: 7 1`] = `
Array [
  "rgba(123, 149, 100, 1)",
  "rgba(120, 145, 93, 1)",
  "rgba(216, 201, 183, 1)",
  "rgba(196, 130, 113, 1)",
  "rgba(226, 211, 195, 1)",
  "rgba(18, 44, 72, 1)",
  "rgba(233, 240, 233, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: default ;numberOfColors: 8 1`] = `
Array [
  "rgba(123, 149, 100, 1)",
  "rgba(120, 145, 93, 1)",
  "rgba(216, 201, 183, 1)",
  "rgba(196, 130, 113, 1)",
  "rgba(226, 211, 194, 1)",
  "rgba(18, 44, 72, 1)",
  "rgba(225, 224, 218, 1)",
  "rgba(234, 243, 235, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: default ;numberOfColors: 27 1`] = `
Array [
  "rgba(116, 141, 93, 1)",
  "rgba(122, 150, 99, 1)",
  "rgba(8, 52, 91, 1)",
  "rgba(217, 203, 191, 1)",
  "rgba(210, 195, 176, 1)",
  "rgba(44, 52, 71, 1)",
  "rgba(17, 62, 100, 1)",
  "rgba(200, 126, 111, 1)",
  "rgba(129, 151, 96, 1)",
  "rgba(219, 203, 182, 1)",
  "rgba(217, 201, 185, 1)",
  "rgba(14, 59, 96, 1)",
  "rgba(246, 253, 247, 1)",
  "rgba(137, 155, 109, 1)",
  "rgba(224, 209, 189, 1)",
  "rgba(214, 208, 204, 1)",
  "rgba(18, 68, 105, 1)",
  "rgba(59, 87, 96, 1)",
  "rgba(224, 216, 206, 1)",
  "rgba(229, 212, 194, 1)",
  "rgba(222, 206, 190, 1)",
  "rgba(17, 13, 23, 1)",
  "rgba(17, 19, 41, 1)",
  "rgba(234, 243, 235, 1)",
  "rgba(232, 216, 198, 1)",
  "rgba(223, 207, 192, 1)",
  "rgba(228, 233, 226, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: default ;numberOfColors: 64 1`] = `
Array [
  "rgba(202, 188, 169, 1)",
  "rgba(216, 220, 212, 1)",
  "rgba(112, 139, 93, 1)",
  "rgba(118, 143, 92, 1)",
  "rgba(120, 146, 95, 1)",
  "rgba(119, 142, 104, 1)",
  "rgba(121, 148, 97, 1)",
  "rgba(11, 56, 94, 1)",
  "rgba(220, 205, 189, 1)",
  "rgba(110, 134, 82, 1)",
  "rgba(215, 199, 181, 1)",
  "rgba(215, 209, 212, 1)",
  "rgba(23, 57, 90, 1)",
  "rgba(118, 148, 95, 1)",
  "rgba(122, 150, 98, 1)",
  "rgba(124, 152, 101, 1)",
  "rgba(16, 66, 102, 1)",
  "rgba(17, 67, 105, 1)",
  "rgba(128, 153, 98, 1)",
  "rgba(128, 150, 94, 1)",
  "rgba(220, 204, 183, 1)",
  "rgba(213, 207, 202, 1)",
  "rgba(16, 61, 100, 1)",
  "rgba(121, 149, 99, 1)",
  "rgba(20, 66, 103, 1)",
  "rgba(32, 73, 98, 1)",
  "rgba(21, 70, 107, 1)",
  "rgba(205, 122, 108, 1)",
  "rgba(85, 97, 95, 1)",
  "rgba(133, 155, 101, 1)",
  "rgba(225, 209, 189, 1)",
  "rgba(222, 224, 218, 1)",
  "rgba(235, 223, 208, 1)",
  "rgba(221, 207, 192, 1)",
  "rgba(19, 68, 106, 1)",
  "rgba(234, 243, 237, 1)",
  "rgba(245, 251, 245, 1)",
  "rgba(15, 10, 17, 1)",
  "rgba(237, 242, 234, 1)",
  "rgba(160, 154, 133, 1)",
  "rgba(225, 210, 194, 1)",
  "rgba(229, 213, 193, 1)",
  "rgba(225, 219, 222, 1)",
  "rgba(222, 206, 189, 1)",
  "rgba(230, 229, 225, 1)",
  "rgba(15, 16, 34, 1)",
  "rgba(232, 241, 234, 1)",
  "rgba(234, 243, 235, 1)",
  "rgba(235, 243, 235, 1)",
  "rgba(231, 241, 232, 1)",
  "rgba(229, 213, 196, 1)",
  "rgba(230, 213, 195, 1)",
  "rgba(227, 229, 222, 1)",
  "rgba(223, 207, 192, 1)",
  "rgba(229, 239, 232, 1)",
  "rgba(16, 19, 42, 1)",
  "rgba(231, 243, 235, 1)",
  "rgba(28, 28, 47, 1)",
  "rgba(14, 18, 44, 1)",
  "rgba(25, 22, 36, 1)",
  "rgba(231, 215, 199, 1)",
  "rgba(230, 213, 195, 1)",
  "rgba(222, 215, 212, 1)",
  "rgba(224, 206, 188, 1)",
]
`;

exports[`palette.ts Palette Palette.imageData colorQuantCycles: default ;numberOfColors: default 1`] = `
Array [
  "rgba(215, 212, 208, 1)",
  "rgba(118, 144, 93, 1)",
  "rgba(124, 151, 99, 1)",
  "rgba(16, 62, 100, 1)",
  "rgba(134, 152, 101, 1)",
  "rgba(222, 222, 217, 1)",
  "rgba(82, 100, 96, 1)",
  "rgba(199, 130, 114, 1)",
  "rgba(233, 242, 235, 1)",
  "rgba(220, 204, 187, 1)",
  "rgba(230, 225, 221, 1)",
  "rgba(27, 67, 99, 1)",
  "rgba(17, 18, 37, 1)",
  "rgba(246, 253, 247, 1)",
  "rgba(229, 213, 195, 1)",
  "rgba(230, 233, 224, 1)",
]
`;

exports[`palette.ts Palette Palette.number numberOfColors: 8 1`] = `
Array [
  "rgba(0, 0, 0, 1)",
  "rgba(0, 0, 255, 1)",
  "rgba(0, 255, 0, 1)",
  "rgba(0, 255, 255, 1)",
  "rgba(255, 0, 0, 1)",
  "rgba(255, 0, 255, 1)",
  "rgba(255, 255, 0, 1)",
  "rgba(255, 255, 255, 1)",
]
`;

exports[`palette.ts Palette Palette.number numberOfColors: 27 1`] = `
Array [
  "rgba(0, 0, 0, 1)",
  "rgba(0, 0, 127, 1)",
  "rgba(0, 0, 254, 1)",
  "rgba(0, 127, 0, 1)",
  "rgba(0, 127, 127, 1)",
  "rgba(0, 127, 254, 1)",
  "rgba(0, 254, 0, 1)",
  "rgba(0, 254, 127, 1)",
  "rgba(0, 254, 254, 1)",
  "rgba(127, 0, 0, 1)",
  "rgba(127, 0, 127, 1)",
  "rgba(127, 0, 254, 1)",
  "rgba(127, 127, 0, 1)",
  "rgba(127, 127, 127, 1)",
  "rgba(127, 127, 254, 1)",
  "rgba(127, 254, 0, 1)",
  "rgba(127, 254, 127, 1)",
  "rgba(127, 254, 254, 1)",
  "rgba(254, 0, 0, 1)",
  "rgba(254, 0, 127, 1)",
  "rgba(254, 0, 254, 1)",
  "rgba(254, 127, 0, 1)",
  "rgba(254, 127, 127, 1)",
  "rgba(254, 127, 254, 1)",
  "rgba(254, 254, 0, 1)",
  "rgba(254, 254, 127, 1)",
  "rgba(254, 254, 254, 1)",
]
`;
